src/comm/Device.cc

```diff
@@ -26,80 +26,175 @@ namespace
+        {
+            CELER_LOG(warning)
+                << "CUDA active device ID unexpectedly changed from "
+                << device.device_id() << " to " << cur_id;
+            device = Device(cur_id);
+        }
+    }
+    return device;
 }
 
 } // namespace
 
+//---------------------------------------------------------------------------//
+// DEVICE FUNCTIONS
 //---------------------------------------------------------------------------//
 /*!
- * Determine whether the CUDA device is enabled.
+ * Get the number of available devices.
  *
- * This is true if and only if CUDA support is built-in, if at least one
+ * This is nonzero if and only if CUDA support is built-in, if at least one
  * CUDA-capable device is present, and if the 'CELER_DISABLE_DEVICE'
  * environment variable is not set.
  */
-bool is_device_enabled()
+unsigned int Device::num_devices()
 {
-    static const bool result = determine_device_enable();
+    static const unsigned int result = determine_num_devices();
     return result;
 }
 
 //---------------------------------------------------------------------------//
 /*!
  * Initialize device in a round-robin fashion from a communicator.
  */
-void initialize_device(const Communicator& comm)
+Device Device::from_round_robin(const Communicator& comm)
+{
+    int num_devices = Device::num_devices();
+    if (!num_devices)
+    {
+        // Null device
+        return {};
+    }
+
+    return Device(comm.rank() % num_devices);
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Construct from a device ID.
+ */
+Device::Device(int id) : id_(id)
+{
+    CELER_EXPECT(id >= 0
+                 && static_cast<unsigned int>(id) < Device::num_devices());
+
+#if CELERITAS_USE_CUDA
+    cudaDeviceProp props;
+    CELER_CUDA_CALL(cudaGetDeviceProperties(&props, id));
+    name_                 = props.name;
+    total_global_mem_     = props.totalGlobalMem;
+    max_threads_          = props.maxThreadsPerMultiProcessor;
+    num_multi_processors_ = props.multiProcessorCount;
+    warp_size_            = props.warpSize;
+#endif
+
+    CELER_ENSURE(*this);
+    CELER_ENSURE(!name_.empty());
+    CELER_ENSURE(total_global_mem_ > 0);
+    CELER_ENSURE(max_threads_ > 0);
+}
+
+//---------------------------------------------------------------------------//
+// FREE FUNCTIONS
+//---------------------------------------------------------------------------//
+/*!
+ * Get the shared default device.
+ */
+const Device& device()
+{
+    return global_device();
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Activate the given device.
+ *
+ * The given device must be set (true result) unless no device has yet been
+ * enabled -- this allows Device::from_round_robin to create "null" devices
+ * when CUDA is disabled.
+ */
+void activate_device(Device&& device)
 {
-    if (!is_device_enabled())
+    CELER_EXPECT(device || !global_device());
```

Comment:
(It probably can not be enforced but) this routine can not be called from multiple threads (because of the static in global_device) and this the converse of `cudaSetDevice` which must be called on all threads (unless the default is wanted, see https://developer.nvidia.com/blog/cuda-pro-tip-always-set-current-device-avoid-multithreading-bugs/)

Reply:
Aha, I hadn't considered the multithreading case -- I assumed our initial implementation would use MPI for CPU parallelism (one process per GPU) as Shift does. @tmdelellis any comments about OpenMPI + threads + CUDA from Shift?